elfi/examples/lotka_volterra.py

```diff
@@ -0,0 +1,253 @@
+        Time allowed for reactions in the Direct method (not the wall time).
+    batch_size : int, optional
+    random_state : np.random.RandomState, optional
+    return_full : bool, optional
+        If True, return a tuple (observed_stock, observed_times, full_stock, full_times).
+
+    Returns
+    -------
+    stock_obs : np.array
+        Observations in shape (batch_size, n_obs, 2).
+
+    """
+    random_state = random_state or np.random
+
+    r1 = np.asanyarray(r1).reshape(-1)
+    r2 = np.asanyarray(r2).reshape(-1)
+    r3 = np.asanyarray(r3).reshape(-1)
+    prey_init = np.asanyarray(prey_init).reshape(-1)
+    predator_init = np.asanyarray(predator_init).reshape(-1)
+    sigma = np.asanyarray(sigma).reshape(-1)
+
+    n_full = 20000
+    stock = np.empty((batch_size, n_full, 2), dtype=np.int32)
+    stock[:, 0, 0] = prey_init
+    stock[:, 0, 1] = predator_init
+    stoichiometry = np.array([[1, 0], [-1, 1], [0, -1], [0, 0]], dtype=np.int32)
+    times = np.empty((batch_size, n_full))
+    times[:, 0] = 0
+    ii = 0
+
+    # iterate until all in batch ok
+    while np.any(times[:, ii] < time_end):
+        ii += 1
+
+        # increase the size of arrays if needed
+        if ii == n_full:
+            stock = np.concatenate((stock, np.empty((batch_size, n_full, 2))), axis=1)
+            times = np.concatenate((times, np.empty((batch_size, n_full))), axis=1)
+            n_full *= 2
+
+        # reaction probabilities
+        hazards = np.column_stack((r1 * stock[:, ii-1, 0],
+                                   r2 * stock[:, ii-1, 0] * stock[:, ii-1, 1],
+                                   r3 * stock[:, ii-1, 1]))
+
+        with np.errstate(divide='ignore', invalid='ignore'):
+            inv_sum_hazards = 1. / np.sum(hazards, axis=1, keepdims=True)  # inf if all dead
+
+            delta_t = random_state.exponential(inv_sum_hazards.ravel())
+            times[:, ii] = times[:, ii-1] + delta_t
+
+            # choose reaction according to their probabilities
+            probs = hazards * inv_sum_hazards
+            cumprobs = np.cumsum(probs[:, :-1], axis=1)
+            x = random_state.uniform(size=(batch_size, 1))
+            reaction = np.sum(x >= cumprobs, axis=1)
+
+        # null reaction if both populations dead
+        reaction = np.where(np.isinf(inv_sum_hazards.ravel()), 3, reaction)
+
+        # update stock
+        stock[:, ii, :] = stock[:, ii-1, :] + stoichiometry[reaction, :]
+
+        # no point to continue if predators = 0
+        times[:, ii] = np.where(stock[:, ii, 1] == 0, time_end, times[:, ii])
+
+    stock = stock[:, :ii+1, :]
+    times = times[:, :ii+1]
+
+    times_out = np.linspace(0, time_end, n_obs)
+    stock_out = np.empty((batch_size, n_obs, 2), dtype=np.int32)
+    stock_out[:, 0, :] = stock[:, 0, :]
+
+    # observations at even intervals
+    for ii in range(1, n_obs):
+        iy, ix = np.where(times >= times_out[ii])
+        iy, iix = np.unique(iy, return_index=True)
```

Comment:
the first `iy` is rewritten by the second one. If the first one is a dummy instance, would suggest calling it `_` instead.